// @appbaseio/reactivesearch/lib/appbase-js/src/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (config) {
  var client = new _index2.default(config);

  _index2.default.prototype.performFetchRequest = _fetch2.default;

  _index2.default.prototype.performWsRequest = _ws2.default;

  _index2.default.prototype.index = _index4.default;

  _index2.default.prototype.get = _get2.default;

  _index2.default.prototype.update = _update2.default;

  _index2.default.prototype.delete = _delete2.default;

  _index2.default.prototype.bulk = _bulk2.default;

  _index2.default.prototype.search = _search2.default;

  _index2.default.prototype.msearch = _msearch2.default;

  _index2.default.prototype.getStream = _getStream2.default;

  _index2.default.prototype.searchStream = _searchStream2.default;

  _index2.default.prototype.searchStreamToURL = _searchStreamToURL2.default;

  _index2.default.prototype.getTypes = _getTypes2.default;

  _index2.default.prototype.getMappings = _getMappings2.default;

  _index2.default.prototype.setHeaders = function () {
    var headers = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var shouldEncode = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

    // Encode headers
    if (shouldEncode) {
      this.headers = (0, _index5.encodeHeaders)(headers);
    } else {
      this.headers = headers;
    }
  };

  if (typeof window !== 'undefined') {
    window.Appbase = client;
  }
  return client;
};

var _index = require('./core/index');

var _index2 = _interopRequireDefault(_index);

var _fetch = require('./core/request/fetch');

var _fetch2 = _interopRequireDefault(_fetch);

var _ws = require('./core/request/ws');

var _ws2 = _interopRequireDefault(_ws);

var _index3 = require('./core/api/index');

var _index4 = _interopRequireDefault(_index3);

var _get = require('./core/api/get');

var _get2 = _interopRequireDefault(_get);

var _update = require('./core/api/update');

var _update2 = _interopRequireDefault(_update);

var _delete = require('./core/api/delete');

var _delete2 = _interopRequireDefault(_delete);

var _bulk = require('./core/api/bulk');

var _bulk2 = _interopRequireDefault(_bulk);

var _search = require('./core/api/search');

var _search2 = _interopRequireDefault(_search);

var _msearch = require('./core/api/msearch');

var _msearch2 = _interopRequireDefault(_msearch);

var _getStream = require('./core/api/getStream');

var _getStream2 = _interopRequireDefault(_getStream);

var _searchStream = require('./core/api/searchStream');

var _searchStream2 = _interopRequireDefault(_searchStream);

var _searchStreamToURL = require('./core/api/searchStreamToURL');

var _searchStreamToURL2 = _interopRequireDefault(_searchStreamToURL);

var _getTypes = require('./core/api/getTypes');

var _getTypes2 = _interopRequireDefault(_getTypes);

var _getMappings = require('./core/api/getMappings');

var _getMappings2 = _interopRequireDefault(_getMappings);

var _index5 = require('./utils/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
