// npm-bundled/index.js
'use strict';

// walk the tree of deps starting from the top level list of bundled deps
// Any deps at the top level that are depended on by a bundled dep that
// does not have that dep in its own node_modules folder are considered
// bundled deps as well.  This list of names can be passed to npm-packlist
// as the "bundled" argument.  Additionally, packageJsonCache is shared so
// packlist doesn't have to re-read files already consumed in this pass

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var fs = require('fs');
var path = require('path');
var EE = require('events').EventEmitter;

var BundleWalker = function (_EE) {
  _inherits(BundleWalker, _EE);

  function BundleWalker(opt) {
    _classCallCheck(this, BundleWalker);

    opt = opt || {};

    var _this = _possibleConstructorReturn(this, (BundleWalker.__proto__ || Object.getPrototypeOf(BundleWalker)).call(this, opt));

    _this.path = path.resolve(opt.path || process.cwd());

    _this.parent = opt.parent || null;
    if (_this.parent) {
      _this.result = _this.parent.result;
      // only collect results in node_modules folders at the top level
      // since the node_modules in a bundled dep is included always
      if (!_this.parent.parent) {
        var base = path.basename(_this.path);
        var scope = path.basename(path.dirname(_this.path));
        _this.result.add(/^@/.test(scope) ? scope + '/' + base : base);
      }
      _this.root = _this.parent.root;
      _this.packageJsonCache = _this.parent.packageJsonCache;
    } else {
      _this.result = new Set();
      _this.root = _this.path;
      _this.packageJsonCache = opt.packageJsonCache || new Map();
    }

    _this.seen = new Set();
    _this.didDone = false;
    _this.children = 0;
    _this.node_modules = [];
    _this.package = null;
    _this.bundle = null;
    return _this;
  }

  _createClass(BundleWalker, [{
    key: 'done',
    value: function done() {
      if (!this.didDone) {
        this.didDone = true;
        if (!this.parent) {
          var res = Array.from(this.result);
          this.result = res;
          this.emit('done', res);
        } else {
          this.emit('done');
        }
      }
    }
  }, {
    key: 'start',
    value: function start() {
      var pj = this.path + '/package.json';
      if (this.packageJsonCache.has(pj)) this.onPackage(this.packageJsonCache.get(pj));else this.readPackageJson(pj);
      return this;
    }
  }, {
    key: 'readPackageJson',
    value: function readPackageJson(pj) {
      var _this2 = this;

      fs.readFile(pj, function (er, data) {
        return er ? _this2.done() : _this2.onPackageJson(pj, data);
      });
    }
  }, {
    key: 'onPackageJson',
    value: function onPackageJson(pj, data) {
      try {
        this.package = JSON.parse(data + '');
      } catch (er) {
        return this.done();
      }
      this.packageJsonCache.set(pj, this.package);
      this.onPackage(this.package);
    }
  }, {
    key: 'onPackage',
    value: function onPackage(pkg) {
      // all deps are bundled if we got here as a child.
      // otherwise, only bundle bundledDeps
      // Get a unique-ified array with a short-lived Set
      var bdRaw = this.parent ? Object.keys(pkg.dependencies || {}).concat(Object.keys(pkg.optionalDependencies || {})) : pkg.bundleDependencies || pkg.bundledDependencies || [];

      var bd = Array.from(new Set(Array.isArray(bdRaw) ? bdRaw : Object.keys(bdRaw)));

      if (!bd.length) return this.done();

      this.bundle = bd;
      var nm = this.path + '/node_modules';
      this.readModules();
    }
  }, {
    key: 'readModules',
    value: function readModules() {
      var _this3 = this;

      readdirNodeModules(this.path + '/node_modules', function (er, nm) {
        return er ? _this3.onReaddir([]) : _this3.onReaddir(nm);
      });
    }
  }, {
    key: 'onReaddir',
    value: function onReaddir(nm) {
      var _this4 = this;

      // keep track of what we have, in case children need it
      this.node_modules = nm;

      this.bundle.forEach(function (dep) {
        return _this4.childDep(dep);
      });
      if (this.children === 0) this.done();
    }
  }, {
    key: 'childDep',
    value: function childDep(dep) {
      if (this.node_modules.indexOf(dep) !== -1 && !this.seen.has(dep)) {
        this.seen.add(dep);
        this.child(dep);
      } else if (this.parent) {
        this.parent.childDep(dep);
      }
    }
  }, {
    key: 'child',
    value: function child(dep) {
      var _this5 = this;

      var p = this.path + '/node_modules/' + dep;
      this.children += 1;
      var child = new BundleWalker({
        path: p,
        parent: this
      });
      child.on('done', function (_) {
        if (--_this5.children === 0) _this5.done();
      });
      child.start();
    }
  }]);

  return BundleWalker;
}(EE);

var BundleWalkerSync = function (_BundleWalker) {
  _inherits(BundleWalkerSync, _BundleWalker);

  function BundleWalkerSync(opt) {
    _classCallCheck(this, BundleWalkerSync);

    return _possibleConstructorReturn(this, (BundleWalkerSync.__proto__ || Object.getPrototypeOf(BundleWalkerSync)).call(this, opt));
  }

  _createClass(BundleWalkerSync, [{
    key: 'start',
    value: function start() {
      _get(BundleWalkerSync.prototype.__proto__ || Object.getPrototypeOf(BundleWalkerSync.prototype), 'start', this).call(this);
      this.done();
      return this;
    }
  }, {
    key: 'readPackageJson',
    value: function readPackageJson(pj) {
      try {
        this.onPackageJson(pj, fs.readFileSync(pj));
      } catch (er) {}
      return this;
    }
  }, {
    key: 'readModules',
    value: function readModules() {
      try {
        this.onReaddir(readdirNodeModulesSync(this.path + '/node_modules'));
      } catch (er) {
        this.onReaddir([]);
      }
    }
  }, {
    key: 'child',
    value: function child(dep) {
      new BundleWalkerSync({
        path: this.path + '/node_modules/' + dep,
        parent: this
      }).start();
    }
  }]);

  return BundleWalkerSync;
}(BundleWalker);

var readdirNodeModules = function readdirNodeModules(nm, cb) {
  fs.readdir(nm, function (er, set) {
    if (er) cb(er);else {
      var scopes = set.filter(function (f) {
        return (/^@/.test(f)
        );
      });
      if (!scopes.length) cb(null, set);else {
        var unscoped = set.filter(function (f) {
          return !/^@/.test(f);
        });
        var count = scopes.length;
        scopes.forEach(function (scope) {
          fs.readdir(nm + '/' + scope, function (er, pkgs) {
            if (er || !pkgs.length) unscoped.push(scope);else unscoped.push.apply(unscoped, pkgs.map(function (p) {
              return scope + '/' + p;
            }));
            if (--count === 0) cb(null, unscoped);
          });
        });
      }
    }
  });
};

var readdirNodeModulesSync = function readdirNodeModulesSync(nm) {
  var set = fs.readdirSync(nm);
  var unscoped = set.filter(function (f) {
    return !/^@/.test(f);
  });
  var scopes = set.filter(function (f) {
    return (/^@/.test(f)
    );
  }).map(function (scope) {
    try {
      var pkgs = fs.readdirSync(nm + '/' + scope);
      return pkgs.length ? pkgs.map(function (p) {
        return scope + '/' + p;
      }) : [scope];
    } catch (er) {
      return [scope];
    }
  }).reduce(function (a, b) {
    return a.concat(b);
  }, []);
  return unscoped.concat(scopes);
};

var walk = function walk(options, callback) {
  var p = new Promise(function (resolve, reject) {
    new BundleWalker(options).on('done', resolve).on('error', reject).start();
  });
  return callback ? p.then(function (res) {
    return callback(null, res);
  }, callback) : p;
};

var walkSync = function walkSync(options) {
  return new BundleWalkerSync(options).start().result;
};

module.exports = walk;
walk.sync = walkSync;
walk.BundleWalker = BundleWalker;
walk.BundleWalkerSync = BundleWalkerSync;
