// uri-js/rollup.config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _rollupPluginNodeResolve = require('rollup-plugin-node-resolve');

var _rollupPluginNodeResolve2 = _interopRequireDefault(_rollupPluginNodeResolve);

var _rollupPluginBabel = require('rollup-plugin-babel');

var _rollupPluginBabel2 = _interopRequireDefault(_rollupPluginBabel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var packageJson = require('./package.json');

exports.default = {
	entry: "dist/esnext/index.js",
	format: "umd",
	moduleName: "URI",
	plugins: [(0, _rollupPluginNodeResolve2.default)({
		module: true,
		jsnext: true,
		preferBuiltins: false
	}), (0, _rollupPluginBabel2.default)({
		"presets": [["latest", {
			"es2015": {
				"modules": false
			}
		}]],
		"plugins": ["external-helpers"],
		"externalHelpers": false
	})],
	dest: "dist/es5/uri.all.js",
	sourceMap: true,
	banner: "/** @license URI.js v" + packageJson.version + " (c) 2011 Gary Court. License: http://github.com/garycourt/uri-js */"
};
