// rollup-plugin-node-builtins/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var E2BIG = exports.E2BIG = 7;
var EACCES = exports.EACCES = 13;
var EADDRINUSE = exports.EADDRINUSE = 48;
var EADDRNOTAVAIL = exports.EADDRNOTAVAIL = 49;
var EAFNOSUPPORT = exports.EAFNOSUPPORT = 47;
var EAGAIN = exports.EAGAIN = 35;
var EALREADY = exports.EALREADY = 37;
var EBADF = exports.EBADF = 9;
var EBADMSG = exports.EBADMSG = 94;
var EBUSY = exports.EBUSY = 16;
var ECANCELED = exports.ECANCELED = 89;
var ECHILD = exports.ECHILD = 10;
var ECONNABORTED = exports.ECONNABORTED = 53;
var ECONNREFUSED = exports.ECONNREFUSED = 61;
var ECONNRESET = exports.ECONNRESET = 54;
var EDEADLK = exports.EDEADLK = 11;
var EDESTADDRREQ = exports.EDESTADDRREQ = 39;
var EDOM = exports.EDOM = 33;
var EDQUOT = exports.EDQUOT = 69;
var EEXIST = exports.EEXIST = 17;
var EFAULT = exports.EFAULT = 14;
var EFBIG = exports.EFBIG = 27;
var EHOSTUNREACH = exports.EHOSTUNREACH = 65;
var EIDRM = exports.EIDRM = 90;
var EILSEQ = exports.EILSEQ = 92;
var EINPROGRESS = exports.EINPROGRESS = 36;
var EINTR = exports.EINTR = 4;
var EINVAL = exports.EINVAL = 22;
var EIO = exports.EIO = 5;
var EISCONN = exports.EISCONN = 56;
var EISDIR = exports.EISDIR = 21;
var ELOOP = exports.ELOOP = 62;
var EMFILE = exports.EMFILE = 24;
var EMLINK = exports.EMLINK = 31;
var EMSGSIZE = exports.EMSGSIZE = 40;
var EMULTIHOP = exports.EMULTIHOP = 95;
var ENAMETOOLONG = exports.ENAMETOOLONG = 63;
var ENETDOWN = exports.ENETDOWN = 50;
var ENETRESET = exports.ENETRESET = 52;
var ENETUNREACH = exports.ENETUNREACH = 51;
var ENFILE = exports.ENFILE = 23;
var ENOBUFS = exports.ENOBUFS = 55;
var ENODATA = exports.ENODATA = 96;
var ENODEV = exports.ENODEV = 19;
var ENOENT = exports.ENOENT = 2;
var ENOEXEC = exports.ENOEXEC = 8;
var ENOLCK = exports.ENOLCK = 77;
var ENOLINK = exports.ENOLINK = 97;
var ENOMEM = exports.ENOMEM = 12;
var ENOMSG = exports.ENOMSG = 91;
var ENOPROTOOPT = exports.ENOPROTOOPT = 42;
var ENOSPC = exports.ENOSPC = 28;
var ENOSR = exports.ENOSR = 98;
var ENOSTR = exports.ENOSTR = 99;
var ENOSYS = exports.ENOSYS = 78;
var ENOTCONN = exports.ENOTCONN = 57;
var ENOTDIR = exports.ENOTDIR = 20;
var ENOTEMPTY = exports.ENOTEMPTY = 66;
var ENOTSOCK = exports.ENOTSOCK = 38;
var ENOTSUP = exports.ENOTSUP = 45;
var ENOTTY = exports.ENOTTY = 25;
var ENXIO = exports.ENXIO = 6;
var EOPNOTSUPP = exports.EOPNOTSUPP = 102;
var EOVERFLOW = exports.EOVERFLOW = 84;
var EPERM = exports.EPERM = 1;
var EPIPE = exports.EPIPE = 32;
var EPROTO = exports.EPROTO = 100;
var EPROTONOSUPPORT = exports.EPROTONOSUPPORT = 43;
var EPROTOTYPE = exports.EPROTOTYPE = 41;
var ERANGE = exports.ERANGE = 34;
var EROFS = exports.EROFS = 30;
var ESPIPE = exports.ESPIPE = 29;
var ESRCH = exports.ESRCH = 3;
var ESTALE = exports.ESTALE = 70;
var ETIME = exports.ETIME = 101;
var ETIMEDOUT = exports.ETIMEDOUT = 60;
var ETXTBSY = exports.ETXTBSY = 26;
var EWOULDBLOCK = exports.EWOULDBLOCK = 35;
var EXDEV = exports.EXDEV = 18;
var SIGHUP = exports.SIGHUP = 1;
var SIGINT = exports.SIGINT = 2;
var SIGQUIT = exports.SIGQUIT = 3;
var SIGILL = exports.SIGILL = 4;
var SIGTRAP = exports.SIGTRAP = 5;
var SIGABRT = exports.SIGABRT = 6;
var SIGIOT = exports.SIGIOT = 6;
var SIGBUS = exports.SIGBUS = 10;
var SIGFPE = exports.SIGFPE = 8;
var SIGKILL = exports.SIGKILL = 9;
var SIGUSR1 = exports.SIGUSR1 = 30;
var SIGSEGV = exports.SIGSEGV = 11;
var SIGUSR2 = exports.SIGUSR2 = 31;
var SIGPIPE = exports.SIGPIPE = 13;
var SIGALRM = exports.SIGALRM = 14;
var SIGTERM = exports.SIGTERM = 15;
var SIGCHLD = exports.SIGCHLD = 20;
var SIGCONT = exports.SIGCONT = 19;
var SIGSTOP = exports.SIGSTOP = 17;
var SIGTSTP = exports.SIGTSTP = 18;
var SIGTTIN = exports.SIGTTIN = 21;
var SIGTTOU = exports.SIGTTOU = 22;
var SIGURG = exports.SIGURG = 16;
var SIGXCPU = exports.SIGXCPU = 24;
var SIGXFSZ = exports.SIGXFSZ = 25;
var SIGVTALRM = exports.SIGVTALRM = 26;
var SIGPROF = exports.SIGPROF = 27;
var SIGWINCH = exports.SIGWINCH = 28;
var SIGIO = exports.SIGIO = 23;
var SIGINFO = exports.SIGINFO = 29;
var SIGSYS = exports.SIGSYS = 12;
var O_RDONLY = exports.O_RDONLY = 0;
var O_WRONLY = exports.O_WRONLY = 1;
var O_RDWR = exports.O_RDWR = 2;
var S_IFMT = exports.S_IFMT = 61440;
var S_IFREG = exports.S_IFREG = 32768;
var S_IFDIR = exports.S_IFDIR = 16384;
var S_IFCHR = exports.S_IFCHR = 8192;
var S_IFBLK = exports.S_IFBLK = 24576;
var S_IFIFO = exports.S_IFIFO = 4096;
var S_IFLNK = exports.S_IFLNK = 40960;
var S_IFSOCK = exports.S_IFSOCK = 49152;
var O_CREAT = exports.O_CREAT = 512;
var O_EXCL = exports.O_EXCL = 2048;
var O_NOCTTY = exports.O_NOCTTY = 131072;
var O_TRUNC = exports.O_TRUNC = 1024;
var O_APPEND = exports.O_APPEND = 8;
var O_DIRECTORY = exports.O_DIRECTORY = 1048576;
var O_NOFOLLOW = exports.O_NOFOLLOW = 256;
var O_SYNC = exports.O_SYNC = 128;
var O_SYMLINK = exports.O_SYMLINK = 2097152;
var O_NONBLOCK = exports.O_NONBLOCK = 4;
var S_IRWXU = exports.S_IRWXU = 448;
var S_IRUSR = exports.S_IRUSR = 256;
var S_IWUSR = exports.S_IWUSR = 128;
var S_IXUSR = exports.S_IXUSR = 64;
var S_IRWXG = exports.S_IRWXG = 56;
var S_IRGRP = exports.S_IRGRP = 32;
var S_IWGRP = exports.S_IWGRP = 16;
var S_IXGRP = exports.S_IXGRP = 8;
var S_IRWXO = exports.S_IRWXO = 7;
var S_IROTH = exports.S_IROTH = 4;
var S_IWOTH = exports.S_IWOTH = 2;
var S_IXOTH = exports.S_IXOTH = 1;
var F_OK = exports.F_OK = 0;
var R_OK = exports.R_OK = 4;
var W_OK = exports.W_OK = 2;
var X_OK = exports.X_OK = 1;
var SSL_OP_ALL = exports.SSL_OP_ALL = 2147486719;
var SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = exports.SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 262144;
var SSL_OP_CIPHER_SERVER_PREFERENCE = exports.SSL_OP_CIPHER_SERVER_PREFERENCE = 4194304;
var SSL_OP_CISCO_ANYCONNECT = exports.SSL_OP_CISCO_ANYCONNECT = 32768;
var SSL_OP_COOKIE_EXCHANGE = exports.SSL_OP_COOKIE_EXCHANGE = 8192;
var SSL_OP_CRYPTOPRO_TLSEXT_BUG = exports.SSL_OP_CRYPTOPRO_TLSEXT_BUG = 2147483648;
var SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS = exports.SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS = 2048;
var SSL_OP_EPHEMERAL_RSA = exports.SSL_OP_EPHEMERAL_RSA = 0;
var SSL_OP_LEGACY_SERVER_CONNECT = exports.SSL_OP_LEGACY_SERVER_CONNECT = 4;
var SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER = exports.SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER = 32;
var SSL_OP_MICROSOFT_SESS_ID_BUG = exports.SSL_OP_MICROSOFT_SESS_ID_BUG = 1;
var SSL_OP_MSIE_SSLV2_RSA_PADDING = exports.SSL_OP_MSIE_SSLV2_RSA_PADDING = 0;
var SSL_OP_NETSCAPE_CA_DN_BUG = exports.SSL_OP_NETSCAPE_CA_DN_BUG = 536870912;
var SSL_OP_NETSCAPE_CHALLENGE_BUG = exports.SSL_OP_NETSCAPE_CHALLENGE_BUG = 2;
var SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG = exports.SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG = 1073741824;
var SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG = exports.SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG = 8;
var SSL_OP_NO_COMPRESSION = exports.SSL_OP_NO_COMPRESSION = 131072;
var SSL_OP_NO_QUERY_MTU = exports.SSL_OP_NO_QUERY_MTU = 4096;
var SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = exports.SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = 65536;
var SSL_OP_NO_SSLv2 = exports.SSL_OP_NO_SSLv2 = 16777216;
var SSL_OP_NO_SSLv3 = exports.SSL_OP_NO_SSLv3 = 33554432;
var SSL_OP_NO_TICKET = exports.SSL_OP_NO_TICKET = 16384;
var SSL_OP_NO_TLSv1 = exports.SSL_OP_NO_TLSv1 = 67108864;
var SSL_OP_NO_TLSv1_1 = exports.SSL_OP_NO_TLSv1_1 = 268435456;
var SSL_OP_NO_TLSv1_2 = exports.SSL_OP_NO_TLSv1_2 = 134217728;
var SSL_OP_PKCS1_CHECK_1 = exports.SSL_OP_PKCS1_CHECK_1 = 0;
var SSL_OP_PKCS1_CHECK_2 = exports.SSL_OP_PKCS1_CHECK_2 = 0;
var SSL_OP_SINGLE_DH_USE = exports.SSL_OP_SINGLE_DH_USE = 1048576;
var SSL_OP_SINGLE_ECDH_USE = exports.SSL_OP_SINGLE_ECDH_USE = 524288;
var SSL_OP_SSLEAY_080_CLIENT_DH_BUG = exports.SSL_OP_SSLEAY_080_CLIENT_DH_BUG = 128;
var SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG = exports.SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG = 0;
var SSL_OP_TLS_BLOCK_PADDING_BUG = exports.SSL_OP_TLS_BLOCK_PADDING_BUG = 512;
var SSL_OP_TLS_D5_BUG = exports.SSL_OP_TLS_D5_BUG = 256;
var SSL_OP_TLS_ROLLBACK_BUG = exports.SSL_OP_TLS_ROLLBACK_BUG = 8388608;
var ENGINE_METHOD_RSA = exports.ENGINE_METHOD_RSA = 1;
var ENGINE_METHOD_DSA = exports.ENGINE_METHOD_DSA = 2;
var ENGINE_METHOD_DH = exports.ENGINE_METHOD_DH = 4;
var ENGINE_METHOD_RAND = exports.ENGINE_METHOD_RAND = 8;
var ENGINE_METHOD_ECDH = exports.ENGINE_METHOD_ECDH = 16;
var ENGINE_METHOD_ECDSA = exports.ENGINE_METHOD_ECDSA = 32;
var ENGINE_METHOD_CIPHERS = exports.ENGINE_METHOD_CIPHERS = 64;
var ENGINE_METHOD_DIGESTS = exports.ENGINE_METHOD_DIGESTS = 128;
var ENGINE_METHOD_STORE = exports.ENGINE_METHOD_STORE = 256;
var ENGINE_METHOD_PKEY_METHS = exports.ENGINE_METHOD_PKEY_METHS = 512;
var ENGINE_METHOD_PKEY_ASN1_METHS = exports.ENGINE_METHOD_PKEY_ASN1_METHS = 1024;
var ENGINE_METHOD_ALL = exports.ENGINE_METHOD_ALL = 65535;
var ENGINE_METHOD_NONE = exports.ENGINE_METHOD_NONE = 0;
var DH_CHECK_P_NOT_SAFE_PRIME = exports.DH_CHECK_P_NOT_SAFE_PRIME = 2;
var DH_CHECK_P_NOT_PRIME = exports.DH_CHECK_P_NOT_PRIME = 1;
var DH_UNABLE_TO_CHECK_GENERATOR = exports.DH_UNABLE_TO_CHECK_GENERATOR = 4;
var DH_NOT_SUITABLE_GENERATOR = exports.DH_NOT_SUITABLE_GENERATOR = 8;
var NPN_ENABLED = exports.NPN_ENABLED = 1;
var ALPN_ENABLED = exports.ALPN_ENABLED = 1;
var RSA_PKCS1_PADDING = exports.RSA_PKCS1_PADDING = 1;
var RSA_SSLV23_PADDING = exports.RSA_SSLV23_PADDING = 2;
var RSA_NO_PADDING = exports.RSA_NO_PADDING = 3;
var RSA_PKCS1_OAEP_PADDING = exports.RSA_PKCS1_OAEP_PADDING = 4;
var RSA_X931_PADDING = exports.RSA_X931_PADDING = 5;
var RSA_PKCS1_PSS_PADDING = exports.RSA_PKCS1_PSS_PADDING = 6;
var POINT_CONVERSION_COMPRESSED = exports.POINT_CONVERSION_COMPRESSED = 2;
var POINT_CONVERSION_UNCOMPRESSED = exports.POINT_CONVERSION_UNCOMPRESSED = 4;
var POINT_CONVERSION_HYBRID = exports.POINT_CONVERSION_HYBRID = 6;
var defaultCoreCipherList = exports.defaultCoreCipherList = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA256:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!SRP:!CAMELLIA";
var defaultCipherList = exports.defaultCipherList = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA256:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!SRP:!CAMELLIA";
exports.default = {
  E2BIG: E2BIG,
  EACCES: EACCES,
  EADDRINUSE: EADDRINUSE,
  EADDRNOTAVAIL: EADDRNOTAVAIL,
  EAFNOSUPPORT: EAFNOSUPPORT,
  EAGAIN: EAGAIN,
  EALREADY: EALREADY,
  EBADF: EBADF,
  EBADMSG: EBADMSG,
  EBUSY: EBUSY,
  ECANCELED: ECANCELED,
  ECHILD: ECHILD,
  ECONNABORTED: ECONNABORTED,
  ECONNREFUSED: ECONNREFUSED,
  ECONNRESET: ECONNRESET,
  EDEADLK: EDEADLK,
  EDESTADDRREQ: EDESTADDRREQ,
  EDOM: EDOM,
  EDQUOT: EDQUOT,
  EEXIST: EEXIST,
  EFAULT: EFAULT,
  EFBIG: EFBIG,
  EHOSTUNREACH: EHOSTUNREACH,
  EIDRM: EIDRM,
  EILSEQ: EILSEQ,
  EINPROGRESS: EINPROGRESS,
  EINTR: EINTR,
  EINVAL: EINVAL,
  EIO: EIO,
  EISCONN: EISCONN,
  EISDIR: EISDIR,
  ELOOP: ELOOP,
  EMFILE: EMFILE,
  EMLINK: EMLINK,
  EMSGSIZE: EMSGSIZE,
  EMULTIHOP: EMULTIHOP,
  ENAMETOOLONG: ENAMETOOLONG,
  ENETDOWN: ENETDOWN,
  ENETRESET: ENETRESET,
  ENETUNREACH: ENETUNREACH,
  ENFILE: ENFILE,
  ENOBUFS: ENOBUFS,
  ENODATA: ENODATA,
  ENODEV: ENODEV,
  ENOENT: ENOENT,
  ENOEXEC: ENOEXEC,
  ENOLCK: ENOLCK,
  ENOLINK: ENOLINK,
  ENOMEM: ENOMEM,
  ENOMSG: ENOMSG,
  ENOPROTOOPT: ENOPROTOOPT,
  ENOSPC: ENOSPC,
  ENOSR: ENOSR,
  ENOSTR: ENOSTR,
  ENOSYS: ENOSYS,
  ENOTCONN: ENOTCONN,
  ENOTDIR: ENOTDIR,
  ENOTEMPTY: ENOTEMPTY,
  ENOTSOCK: ENOTSOCK,
  ENOTSUP: ENOTSUP,
  ENOTTY: ENOTTY,
  ENXIO: ENXIO,
  EOPNOTSUPP: EOPNOTSUPP,
  EOVERFLOW: EOVERFLOW,
  EPERM: EPERM,
  EPIPE: EPIPE,
  EPROTO: EPROTO,
  EPROTONOSUPPORT: EPROTONOSUPPORT,
  EPROTOTYPE: EPROTOTYPE,
  ERANGE: ERANGE,
  EROFS: EROFS,
  ESPIPE: ESPIPE,
  ESRCH: ESRCH,
  ESTALE: ESTALE,
  ETIME: ETIME,
  ETIMEDOUT: ETIMEDOUT,
  ETXTBSY: ETXTBSY,
  EWOULDBLOCK: EWOULDBLOCK,
  EXDEV: EXDEV,
  SIGHUP: SIGHUP,
  SIGINT: SIGINT,
  SIGQUIT: SIGQUIT,
  SIGILL: SIGILL,
  SIGTRAP: SIGTRAP,
  SIGABRT: SIGABRT,
  SIGIOT: SIGIOT,
  SIGBUS: SIGBUS,
  SIGFPE: SIGFPE,
  SIGKILL: SIGKILL,
  SIGUSR1: SIGUSR1,
  SIGSEGV: SIGSEGV,
  SIGUSR2: SIGUSR2,
  SIGPIPE: SIGPIPE,
  SIGALRM: SIGALRM,
  SIGTERM: SIGTERM,
  SIGCHLD: SIGCHLD,
  SIGCONT: SIGCONT,
  SIGSTOP: SIGSTOP,
  SIGTSTP: SIGTSTP,
  SIGTTIN: SIGTTIN,
  SIGTTOU: SIGTTOU,
  SIGURG: SIGURG,
  SIGXCPU: SIGXCPU,
  SIGXFSZ: SIGXFSZ,
  SIGVTALRM: SIGVTALRM,
  SIGPROF: SIGPROF,
  SIGWINCH: SIGWINCH,
  SIGIO: SIGIO,
  SIGINFO: SIGINFO,
  SIGSYS: SIGSYS,
  O_RDONLY: O_RDONLY,
  O_WRONLY: O_WRONLY,
  O_RDWR: O_RDWR,
  S_IFMT: S_IFMT,
  S_IFREG: S_IFREG,
  S_IFDIR: S_IFDIR,
  S_IFCHR: S_IFCHR,
  S_IFBLK: S_IFBLK,
  S_IFIFO: S_IFIFO,
  S_IFLNK: S_IFLNK,
  S_IFSOCK: S_IFSOCK,
  O_CREAT: O_CREAT,
  O_EXCL: O_EXCL,
  O_NOCTTY: O_NOCTTY,
  O_TRUNC: O_TRUNC,
  O_APPEND: O_APPEND,
  O_DIRECTORY: O_DIRECTORY,
  O_NOFOLLOW: O_NOFOLLOW,
  O_SYNC: O_SYNC,
  O_SYMLINK: O_SYMLINK,
  O_NONBLOCK: O_NONBLOCK,
  S_IRWXU: S_IRWXU,
  S_IRUSR: S_IRUSR,
  S_IWUSR: S_IWUSR,
  S_IXUSR: S_IXUSR,
  S_IRWXG: S_IRWXG,
  S_IRGRP: S_IRGRP,
  S_IWGRP: S_IWGRP,
  S_IXGRP: S_IXGRP,
  S_IRWXO: S_IRWXO,
  S_IROTH: S_IROTH,
  S_IWOTH: S_IWOTH,
  S_IXOTH: S_IXOTH,
  F_OK: F_OK,
  R_OK: R_OK,
  W_OK: W_OK,
  X_OK: X_OK,
  SSL_OP_ALL: SSL_OP_ALL,
  SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION: SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION,
  SSL_OP_CIPHER_SERVER_PREFERENCE: SSL_OP_CIPHER_SERVER_PREFERENCE,
  SSL_OP_CISCO_ANYCONNECT: SSL_OP_CISCO_ANYCONNECT,
  SSL_OP_COOKIE_EXCHANGE: SSL_OP_COOKIE_EXCHANGE,
  SSL_OP_CRYPTOPRO_TLSEXT_BUG: SSL_OP_CRYPTOPRO_TLSEXT_BUG,
  SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS: SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS,
  SSL_OP_EPHEMERAL_RSA: SSL_OP_EPHEMERAL_RSA,
  SSL_OP_LEGACY_SERVER_CONNECT: SSL_OP_LEGACY_SERVER_CONNECT,
  SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER: SSL_OP_MICROSOFT_BIG_SSLV3_BUFFER,
  SSL_OP_MICROSOFT_SESS_ID_BUG: SSL_OP_MICROSOFT_SESS_ID_BUG,
  SSL_OP_MSIE_SSLV2_RSA_PADDING: SSL_OP_MSIE_SSLV2_RSA_PADDING,
  SSL_OP_NETSCAPE_CA_DN_BUG: SSL_OP_NETSCAPE_CA_DN_BUG,
  SSL_OP_NETSCAPE_CHALLENGE_BUG: SSL_OP_NETSCAPE_CHALLENGE_BUG,
  SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG: SSL_OP_NETSCAPE_DEMO_CIPHER_CHANGE_BUG,
  SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG: SSL_OP_NETSCAPE_REUSE_CIPHER_CHANGE_BUG,
  SSL_OP_NO_COMPRESSION: SSL_OP_NO_COMPRESSION,
  SSL_OP_NO_QUERY_MTU: SSL_OP_NO_QUERY_MTU,
  SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION: SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION,
  SSL_OP_NO_SSLv2: SSL_OP_NO_SSLv2,
  SSL_OP_NO_SSLv3: SSL_OP_NO_SSLv3,
  SSL_OP_NO_TICKET: SSL_OP_NO_TICKET,
  SSL_OP_NO_TLSv1: SSL_OP_NO_TLSv1,
  SSL_OP_NO_TLSv1_1: SSL_OP_NO_TLSv1_1,
  SSL_OP_NO_TLSv1_2: SSL_OP_NO_TLSv1_2,
  SSL_OP_PKCS1_CHECK_1: SSL_OP_PKCS1_CHECK_1,
  SSL_OP_PKCS1_CHECK_2: SSL_OP_PKCS1_CHECK_2,
  SSL_OP_SINGLE_DH_USE: SSL_OP_SINGLE_DH_USE,
  SSL_OP_SINGLE_ECDH_USE: SSL_OP_SINGLE_ECDH_USE,
  SSL_OP_SSLEAY_080_CLIENT_DH_BUG: SSL_OP_SSLEAY_080_CLIENT_DH_BUG,
  SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG: SSL_OP_SSLREF2_REUSE_CERT_TYPE_BUG,
  SSL_OP_TLS_BLOCK_PADDING_BUG: SSL_OP_TLS_BLOCK_PADDING_BUG,
  SSL_OP_TLS_D5_BUG: SSL_OP_TLS_D5_BUG,
  SSL_OP_TLS_ROLLBACK_BUG: SSL_OP_TLS_ROLLBACK_BUG,
  ENGINE_METHOD_RSA: ENGINE_METHOD_RSA,
  ENGINE_METHOD_DSA: ENGINE_METHOD_DSA,
  ENGINE_METHOD_DH: ENGINE_METHOD_DH,
  ENGINE_METHOD_RAND: ENGINE_METHOD_RAND,
  ENGINE_METHOD_ECDH: ENGINE_METHOD_ECDH,
  ENGINE_METHOD_ECDSA: ENGINE_METHOD_ECDSA,
  ENGINE_METHOD_CIPHERS: ENGINE_METHOD_CIPHERS,
  ENGINE_METHOD_DIGESTS: ENGINE_METHOD_DIGESTS,
  ENGINE_METHOD_STORE: ENGINE_METHOD_STORE,
  ENGINE_METHOD_PKEY_METHS: ENGINE_METHOD_PKEY_METHS,
  ENGINE_METHOD_PKEY_ASN1_METHS: ENGINE_METHOD_PKEY_ASN1_METHS,
  ENGINE_METHOD_ALL: ENGINE_METHOD_ALL,
  ENGINE_METHOD_NONE: ENGINE_METHOD_NONE,
  DH_CHECK_P_NOT_SAFE_PRIME: DH_CHECK_P_NOT_SAFE_PRIME,
  DH_CHECK_P_NOT_PRIME: DH_CHECK_P_NOT_PRIME,
  DH_UNABLE_TO_CHECK_GENERATOR: DH_UNABLE_TO_CHECK_GENERATOR,
  DH_NOT_SUITABLE_GENERATOR: DH_NOT_SUITABLE_GENERATOR,
  NPN_ENABLED: NPN_ENABLED,
  ALPN_ENABLED: ALPN_ENABLED,
  RSA_PKCS1_PADDING: RSA_PKCS1_PADDING,
  RSA_SSLV23_PADDING: RSA_SSLV23_PADDING,
  RSA_NO_PADDING: RSA_NO_PADDING,
  RSA_PKCS1_OAEP_PADDING: RSA_PKCS1_OAEP_PADDING,
  RSA_X931_PADDING: RSA_X931_PADDING,
  RSA_PKCS1_PSS_PADDING: RSA_PKCS1_PSS_PADDING,
  POINT_CONVERSION_COMPRESSED: POINT_CONVERSION_COMPRESSED,
  POINT_CONVERSION_UNCOMPRESSED: POINT_CONVERSION_UNCOMPRESSED,
  POINT_CONVERSION_HYBRID: POINT_CONVERSION_HYBRID,
  defaultCoreCipherList: defaultCoreCipherList,
  defaultCipherList: defaultCipherList
};
