// jest-util/build/Console.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _assert;

function _load_assert() {
  return _assert = _interopRequireDefault(require('assert'));
}

var _util;

function _load_util() {
  return _util = require('util');
}

var _console;

function _load_console() {
  return _console = require('console');
}

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _clear_line;

function _load_clear_line() {
  return _clear_line = _interopRequireDefault(require('./clear_line'));
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 */
/* global stream$Writable */

var CustomConsole = function (_Console) {
  _inherits(CustomConsole, _Console);

  function CustomConsole(stdout, stderr, formatBuffer) {
    _classCallCheck(this, CustomConsole);

    var _this = _possibleConstructorReturn(this, (CustomConsole.__proto__ || Object.getPrototypeOf(CustomConsole)).call(this, stdout, stderr));

    _this._formatBuffer = formatBuffer || function (type, message) {
      return message;
    };
    _this._counters = {};
    _this._timers = {};
    _this._groupDepth = 0;
    return _this;
  }

  _createClass(CustomConsole, [{
    key: '_logToParentConsole',
    value: function _logToParentConsole(message) {
      _get(CustomConsole.prototype.__proto__ || Object.getPrototypeOf(CustomConsole.prototype), 'log', this).call(this, message);
    }
  }, {
    key: '_log',
    value: function _log(type, message) {
      (0, (_clear_line || _load_clear_line()).default)(this._stdout);
      this._logToParentConsole(this._formatBuffer(type, '  '.repeat(this._groupDepth) + message));
    }
  }, {
    key: 'assert',
    value: function assert() {
      try {
        (_assert || _load_assert()).default.apply(undefined, arguments);
      } catch (error) {
        this._log('assert', error.toString());
      }
    }
  }, {
    key: 'count',
    value: function count() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      if (!this._counters[label]) {
        this._counters[label] = 0;
      }

      this._log('count', (0, (_util || _load_util()).format)(label + ': ' + ++this._counters[label]));
    }
  }, {
    key: 'countReset',
    value: function countReset() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      this._counters[label] = 0;
    }
  }, {
    key: 'debug',
    value: function debug() {
      this._log('debug', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'dir',
    value: function dir() {
      this._log('dir', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'dirxml',
    value: function dirxml() {
      this._log('dirxml', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'error',
    value: function error() {
      this._log('error', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'group',
    value: function group() {
      this._groupDepth++;

      if (arguments.length > 0) {
        this._log('group', (_chalk || _load_chalk()).default.bold((_util || _load_util()).format.apply(undefined, arguments)));
      }
    }
  }, {
    key: 'groupCollapsed',
    value: function groupCollapsed() {
      this._groupDepth++;

      if (arguments.length > 0) {
        this._log('groupCollapsed', (_chalk || _load_chalk()).default.bold((_util || _load_util()).format.apply(undefined, arguments)));
      }
    }
  }, {
    key: 'groupEnd',
    value: function groupEnd() {
      if (this._groupDepth > 0) {
        this._groupDepth--;
      }
    }
  }, {
    key: 'info',
    value: function info() {
      this._log('info', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'log',
    value: function log() {
      this._log('log', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'time',
    value: function time() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      if (this._timers[label]) {
        return;
      }

      this._timers[label] = new Date();
    }
  }, {
    key: 'timeEnd',
    value: function timeEnd() {
      var label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';

      var startTime = this._timers[label];

      if (startTime) {
        var endTime = new Date();
        var time = endTime - startTime;
        this._log('time', (0, (_util || _load_util()).format)(label + ': ' + time + 'ms'));
        delete this._timers[label];
      }
    }
  }, {
    key: 'warn',
    value: function warn() {
      this._log('warn', (_util || _load_util()).format.apply(undefined, arguments));
    }
  }, {
    key: 'getBuffer',
    value: function getBuffer() {
      return null;
    }
  }]);

  return CustomConsole;
}((_console || _load_console()).Console);

exports.default = CustomConsole;
