// log-update/index.js
'use strict';

var ansiEscapes = require('ansi-escapes');
var cliCursor = require('cli-cursor');
var wrapAnsi = require('wrap-ansi');

var getWidth = function getWidth(stream) {
	var columns = stream.columns;

	if (!columns) {
		return 80;
	}

	// Windows appears to wrap a character early
	// I hate Windows so much
	if (process.platform === 'win32') {
		return columns - 1;
	}

	return columns;
};

var main = function main(stream, options) {
	options = Object.assign({
		showCursor: false
	}, options);

	var prevLineCount = 0;

	var render = function render() {
		if (!options.showCursor) {
			cliCursor.hide();
		}

		var out = [].join.call(arguments, ' ') + '\n';
		out = wrapAnsi(out, getWidth(stream), {
			trim: false,
			hard: true,
			wordWrap: false
		});
		stream.write(ansiEscapes.eraseLines(prevLineCount) + out);
		prevLineCount = out.split('\n').length;
	};

	render.clear = function () {
		stream.write(ansiEscapes.eraseLines(prevLineCount));
		prevLineCount = 0;
	};

	render.done = function () {
		prevLineCount = 0;

		if (!options.showCursor) {
			cliCursor.show();
		}
	};

	return render;
};

module.exports = main(process.stdout);
module.exports.stderr = main(process.stderr);
module.exports.create = main;
