// lint-staged/src/makeCmdTasks.js
'use strict';

var resolveTaskFn = require('./resolveTaskFn');
var resolveGitDir = require('./resolveGitDir');

var debug = require('debug')('lint-staged:make-cmd-tasks');

/**
 * Creates and returns an array of listr tasks which map to the given commands.
 *
 * @param {Array<string>|string} commands
 * @param {Array<string>} pathsToLint
 * @param {Object} [options]
 * @param {number} options.chunkSize
 * @param {number} options.subTaskConcurrency
 */
module.exports = function makeCmdTasks(commands, pathsToLint) {
  var _ref = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref$chunkSize = _ref.chunkSize,
      chunkSize = _ref$chunkSize === undefined ? Number.MAX_SAFE_INTEGER : _ref$chunkSize,
      _ref$subTaskConcurren = _ref.subTaskConcurrency,
      subTaskConcurrency = _ref$subTaskConcurren === undefined ? 1 : _ref$subTaskConcurren;

  debug('Creating listr tasks for commands %o', commands);

  var gitDir = resolveGitDir();
  var lintersArray = Array.isArray(commands) ? commands : [commands];

  return lintersArray.map(function (linter) {
    return {
      title: linter,
      task: resolveTaskFn({
        linter: linter,
        gitDir: gitDir,
        pathsToLint: pathsToLint,
        chunkSize: chunkSize,
        subTaskConcurrency: subTaskConcurrency
      })
    };
  });
};
